canals/component/component.py

```diff
@@ -134,10 +134,22 @@ def __call__(cls, *args, **kwargs):
             instance.__canals_output__ = deepcopy(getattr(instance.run, "_output_types_cache", {}))
 
         # If the __init__ called component.set_input_types(), __canals_input__ is already populated
-        if not hasattr(instance, "__canals_input__"):
-            run_signature = inspect.signature(getattr(cls, "run"))
+        # and we only need to check for defaults
+        run_signature = inspect.signature(getattr(cls, "run"))
+        if hasattr(instance, "__canals_input__"):
+            for param in list(run_signature.parameters)[1:]:  # First is 'self' and it doesn't matter.
+                if run_signature.parameters[param].default != inspect.Parameter.empty:
+                    if param not in instance.__canals_input__:
+                        instance.__canals_input__[param] = InputSocket(
+                            name=param,
+                            type=run_signature.parameters[param].annotation,
+                            is_mandatory=run_signature.parameters[param].default == inspect.Parameter.empty,
+                        )
+                    else:
+                        instance.__canals_input__[param].is_mandatory = False
```

Comment:
Clarification needed - don't we also need to check on loc 149 if it is mandatory or not using `run_signature.parameters[param].default == inspect.Parameter.empty`. Help me understand this :-)

Reply:
There are two situations possible:

- in one case, the run() method contains sockets that were not defined by `set_input_types`. In this case, the sockets are created as usual and added to `__canals_inputs__`.
- in the other case, the socket was already created with `set_input_types`, and the `run()` method is only used to add a default to it. In that case, we only change the value of `is_mandatory` to `False`.

As I write this, I realize that there is a slightly easier way to write it. Let me simplify it and see if it's more readable!